oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/migration/NodeStateCopier.java

```diff
@@ -145,29 +150,73 @@ public static boolean copyNodeStore(@NotNull final NodeStore source, @NotNull fi
      *
      * @param source The NodeState to copy from.
      * @param target The NodeBuilder to copy to.
+     * @param preserveOnTarget boolean to indicate no changes on target except additions
+     * @param path current path
      * @return Whether changes were made or not.
      */
-    public static boolean copyProperties(NodeState source, NodeBuilder target) {
+    public static boolean copyProperties(NodeState source, NodeBuilder target, boolean preserveOnTarget, String path) {
         boolean hasChanges = false;
 
         // remove removed properties
-        for (final PropertyState property : target.getProperties()) {
-            final String name = property.getName();
-            if (!source.hasProperty(name)) {
-                target.removeProperty(name);
-                hasChanges = true;
+        if (!preserveOnTarget) {
+            for (final PropertyState property : target.getProperties()) {
+                final String name = property.getName();
+                if (!source.hasProperty(name)) {
+                    target.removeProperty(name);
+                    hasChanges = true;
+                }
             }
         }
 
         // add new properties and change changed properties
         for (PropertyState property : source.getProperties()) {
             if (!property.equals(target.getProperty(property.getName()))) {
-                target.setProperty(property);
-                hasChanges = true;
+                if (!isVersionPropertyEmpty(source, property, preserveOnTarget, path)) {
+                    target.setProperty(property);
+                    hasChanges = true;
+                }
             }
         }
         return hasChanges;
     }
+    
+    private static Set<String> getValues(NodeState nodeState, String prop) {
+        if (nodeState.hasProperty(prop)) {
+            Iterable<String> values = nodeState.getProperty(prop).getValue(Type.STRINGS);
+            return StreamSupport.stream(values.spliterator(), false).filter(s -> !s.isEmpty()).collect(Collectors.toSet());
+        }
+        return new HashSet<>();
```

Comment:
Implementation of this method only uses the size of `set` returned by this method. So instead of creating an `set` of non empty properties, we should simply return a boolean indicating whether it is empty or not.
```suggestion
    private static boolean isVersionEmpty(NodeState nodeState, String prop) {
        if (nodeState.hasProperty(prop)) {
            Iterable<String> values = nodeState.getProperty(prop).getValue(Type.STRINGS);
            return StreamSupport.stream(values.spliterator(), false).allMatch(String::isEmpty);
        }
        return true;
```

Reply:
I had left it to leave room for furture in case needed but not a strong opinion here.